tiledb/fragment.cc

```diff
@@ -0,0 +1,215 @@
+#include <pybind11/pybind11.h>
+#include <pybind11/pytypes.h>
+#include <pybind11/numpy.h>
+#include <pybind11/stl.h>
+
+#include <exception>
+
+#define TILEDB_DEPRECATED
+#define TILEDB_DEPRECATED_EXPORT
+
+#include <tiledb/tiledb> // C++
+#include "util.h"
+
+#if !defined(NDEBUG)
+#include "debug.cc"
+#endif
+
+namespace tiledbpy {
+
+using namespace std;
+using namespace tiledb;
+namespace py = pybind11;
+using namespace pybind11::literals;
+
+class PyFragmentInfo {
+
+private:
+    Context ctx_;
+    shared_ptr<tiledb::FragmentInfo> fi_;
+
+public:
+    tiledb_ctx_t *c_ctx_;
+
+public:
+    PyFragmentInfo() = delete;
+
+    PyFragmentInfo(py::object ctx, const string& uri) {
+        tiledb_ctx_t *c_ctx_ = (py::capsule)ctx.attr("__capsule__")();
+        if (c_ctx_ == nullptr)
+            TPY_ERROR_LOC("Invalid context pointer!");
+        ctx_ = Context(c_ctx_, false);
+
+        fi_ = shared_ptr<tiledb::FragmentInfo>(new FragmentInfo(ctx_, uri));
+    }
+
+    void load() const {
+        try{
+            fi_->load();
+        } catch(TileDBError& e) {
+            TPY_ERROR_LOC(e.what());
+        }
+    }
+
+    void load(tiledb_encryption_type_t encryption_type,
+              const string& encryption_key) const {
+        try{
+            fi_->load(encryption_type, encryption_key);
+        } catch(TileDBError& e) {
+            TPY_ERROR_LOC(e.what());
+        }
+    }
+
+    string fragment_uri(uint32_t fid) const {
+        return fi_->fragment_uri(fid);
+    }
+
+    py::tuple get_non_empty_domain(uint32_t fid, uint32_t did, py::dtype type) const {   
+        py::dtype array_type = type.kind() == 'M' ? py::dtype::of<uint64_t>() : type;
+        py::array domain = py::array(array_type, 2);
+        py::buffer_info buffer = domain.request();
+        fi_->get_non_empty_domain(fid, did, buffer.ptr);
+
+        if(type.kind() == 'M'){
+            auto np = py::module::import("numpy");
+            auto datetime64 = np.attr("datetime64");
+            auto datetime_data = np.attr("datetime_data");
+
+            uint64_t* dates = static_cast<uint64_t*>(buffer.ptr);
+            domain = py::make_tuple(datetime64(dates[0], datetime_data(type)), 
+                                    datetime64(dates[1], datetime_data(type)));
+        }
+
+        return domain;
+    }
+
+    py::tuple get_non_empty_domain(uint32_t fid, const string& dim_name, py::dtype type) const {
```

Comment:
Minor point, for potential follow-up PR: this is slightly repetitive -- could get the dimension index from the name here, and then call the signature above.